File: src-model/model/celestials/parts/Clouds.java
package model.celestials.parts;

import model.celestials.CelestialBody;
import model.interfaces.Drawable;
import model.type.DrawableResolution;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.jogamp.opengl.util.texture.Texture;

/**
 * Object which stores the clouds parameters for {@link Sphere} part of {@link CelestialBody} entity.
 */
public class Clouds implements Drawable {

    private String textureFile;
    private double radius;

    /* ========== PUBLIC ========== */
    public Clouds(String textureFile, double radius) {
        this.textureFile = textureFile;
        this.radius = radius;
        
        this.resolution = DrawableResolution.VERY_FAR;
    }

    /* ========== PRIVATE ========== */
    @JsonIgnore
    private Texture texture;
    @JsonIgnore
    private DrawableResolution resolution;

    /* ========== PROPERTIES ========== */
    public String getTextureFile() {
        return textureFile;
    }

    public void setTextureFile(String textureFile) {
        this.textureFile = textureFile;
    }

    public double getRadius() {
        return radius;
    }

    public void setRadius(double radius) {
        this.radius = radius;
    }

    public Texture getTexture() {
        return texture;
    }

    public void setTexture(Texture texture) {
        this.texture = texture;
    }

    public DrawableResolution getResolution() {
        return resolution;
    }

    public void setResolution(DrawableResolution resolution) {
        this.resolution = resolution;
    }
}

File: src-model/model/celestials/parts/Clouds.kt
package model.celestials.parts

import com.fasterxml.jackson.annotation.JsonIgnore
import engine.components.drawers.DrawableClouds
import engine.math.Radius
import engine.utils.TextureLoader
import model.celestials.CelestialBody
import model.interfaces.Drawable
import model.type.DrawableResolution
import javax.media.opengl.GL2

/**
 * Object which stores the clouds parameters for [Sphere] part of [CelestialBody] entity.
 */
open class Clouds(
        var textureFile: String?,
        var radius: Radius
) : Drawable {
    @JsonIgnore
    var resolution: DrawableResolution = DrawableResolution.VERY_FAR

    @JsonIgnore
    var drawable : DrawableClouds? = null

    override fun draw(gl: GL2) {
        if (this.drawable == null) {
            this.drawable = DrawableClouds(this, TextureLoader.getTexture(gl, textureFile))
        }

        this.drawable?.draw(gl)
    }
}